File: viewer/src/main/scala/farjs/viewer/ViewerContent.scala
package farjs.viewer

import farjs.file.popups._
import farjs.filelist.theme.FileListTheme
import scommons.react._
import scommons.react.blessed._
import scommons.react.hooks._

import scala.scalajs.js

object ViewerContent extends FunctionComponent[ViewerContentProps] {
  
  private[viewer] var viewerInput: ReactClass = ViewerInput
  private[viewer] var encodingsPopup: ReactClass = EncodingsPopup
  private[viewer] var textSearchPopup: ReactClass = TextSearchPopup
  private[viewer] var viewerSearch: ReactClass = ViewerSearch

  protected def render(compProps: Props): ReactElement = {
    val theme = FileListTheme.useTheme()
    val props = compProps.plain
    val viewport = props.viewport
    val readF = useRef(js.Promise.resolve[ViewerFileViewport](viewport))
    val (showEncodingsPopup, setShowEncodingsPopup) = useState(false)
    val (showSearchPopup, setShowSearchPopup) = useState(false)
    val (searchTerm, setSearchTerm) = useState("")
    
    val updated: js.Function1[ViewerFileViewport, js.Promise[ViewerFileViewport]] = { viewport =>
      props.setViewport(viewport)
      js.Promise.resolve[ViewerFileViewport](viewport)
    }
    
    def onMoveUp(lines: Int, from: Double = viewport.position): Unit = {
      readF.current = readF.current.`then`[ViewerFileViewport](
        (viewport => viewport.moveUp(lines, from)): js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]]
      ).`then`[ViewerFileViewport](updated)
    }
    
    def onMoveDown(lines: Int): Unit = {
      readF.current = readF.current.`then`[ViewerFileViewport](
        (viewport => viewport.moveDown(lines)): js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]]
      ).`then`[ViewerFileViewport](updated)
    }
    
    def onReload(from: Double = viewport.position): Unit = {
      readF.current = readF.current.`then`[ViewerFileViewport](
        (viewport => viewport.reload(from)): js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]]
      ).`then`[ViewerFileViewport](updated)
    }
    
    def onWrap(): Unit = {
      readF.current = readF.current.`then`[ViewerFileViewport]({ viewport =>
        val wrap = !viewport.wrap
        val column =
          if (wrap) 0
          else viewport.column

        val newWrap = wrap
        val newColumn = column
        updated(viewport.updated(new ViewerFileViewportData {
          override val wrap = newWrap
          override val column = newColumn
        }))
      }: js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]])
    }
    
    val onEncoding: js.Function1[String, Unit] = { newEncoding =>
      readF.current = readF.current.`then`[ViewerFileViewport]({ viewport =>
        updated(viewport.updated(new ViewerFileViewportData {
          override val encoding = newEncoding
        }))
      }: js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]])
    }
    
    def onColumn(dx: Int): Unit = {
      readF.current = readF.current.`then`[ViewerFileViewport]({ viewport =>
        val col = viewport.column + dx
        if (col >= 0 && col < 1000) {
          updated(viewport.updated(new ViewerFileViewportData {
            override val column = col
          }))
        }
        else js.Promise.resolve[ViewerFileViewport](viewport)
      }: js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]])
    }
    
    val onKeypress: js.Function1[String, Unit] = { keyFull =>
      if (!props.onKeypress(keyFull)) {
        keyFull match {
          case "f2" => onWrap()
          case "f7" => setShowSearchPopup(true)
          case "f8" => setShowEncodingsPopup(true)
          case "left" => onColumn(dx = -1)
          case "right" => onColumn(dx = 1)
          case "C-r" => onReload()
          case "home" => onReload(from = 0)
          case "end" => onMoveUp(lines = viewport.height, from = viewport.size)
          case "up" => onMoveUp(lines = 1)
          case "pageup" => onMoveUp(lines = viewport.height)
          case "down" => onMoveDown(lines = 1)
          case "pagedown" => onMoveDown(lines = viewport.height)
          case _ =>
        }
      }
    }
    
    useLayoutEffect({ () =>
      readF.current = readF.current.`then`[ViewerFileViewport](
        (_ => viewport.reload()): js.Function1[ViewerFileViewport, js.Thenable[ViewerFileViewport]]
      ).`then`[ViewerFileViewport](updated)
      ()
    }, List(viewport.encoding, viewport.size, viewport.width, viewport.height, viewport.wrap))

    <(viewerInput)(^.plain := ViewerInputProps(
      inputRef = props.inputRef,
      onWheel = { up =>
        if (up) onMoveUp(lines = 1)
        else onMoveDown(lines = 1)
      },
      onKeypress = onKeypress
    ))(
      <.text(
        ^.rbWidth := viewport.width,
        ^.rbHeight := viewport.height,
        ^.rbStyle := ViewerController.contentStyle(theme),
        ^.rbWrap := false,
        ^.content := viewport.content
      )(),

      if (showEncodingsPopup) Some {
        <(encodingsPopup)(^.plain := EncodingsPopupProps(
          encoding = props.viewport.encoding,
          onApply = onEncoding,
          onClose = { () =>
            setShowEncodingsPopup(false)
          }
        ))()
      }
      else None,

      if (showSearchPopup) Some {
        <(textSearchPopup)(^.plain := TextSearchPopupProps(
          onSearch = { term =>
            setShowSearchPopup(false)
            setSearchTerm(term)
          },
          onCancel = { () =>
            setShowSearchPopup(false)
          }
        ))()
      }
      else None,

      if (searchTerm.nonEmpty) Some {
        <(viewerSearch)(^.plain := ViewerSearchProps(
          searchTerm = searchTerm,
          onComplete = { () =>
            setSearchTerm("")
          }
        ))()
      }
      else None
    )
  }
}

File: viewer/ViewerContent.mjs
/**
 * @typedef {import("@farjs/blessed").Widgets.BlessedElement} BlessedElement
 * @typedef {import("@farjs/blessed").Widgets.Types.TStyle} BlessedStyle
 * @typedef {import("@farjs/filelist/theme/FileListTheme.mjs").FileListTheme} FileListTheme
 * @import { ViewerFileViewport } from "./ViewerFileViewport.mjs"
 */
import React, { useLayoutEffect, useRef, useState } from "react";
import FileListTheme from "@farjs/filelist/theme/FileListTheme.mjs";
import EncodingsPopup from "../file/popups/EncodingsPopup.mjs";
import TextSearchPopup from "../file/popups/TextSearchPopup.mjs";
import ViewerInput from "./ViewerInput.mjs";
import ViewerSearch from "./ViewerSearch.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly inputRef: React.MutableRefObject<BlessedElement | null>;
 *  readonly viewport: ViewerFileViewport;
 *  setViewport(viewport: ViewerFileViewport | undefined): void;
 *  onKeypress(keyFull: string): boolean;
 * }} ViewerContentProps
 */

/**
 * @param {ViewerContentProps} props
 */
const ViewerContent = (props) => {
  const { viewerInput, encodingsPopup, textSearchPopup, viewerSearch } =
    ViewerContent;

  const theme = FileListTheme.useTheme();
  const viewport = props.viewport;
  const readP = useRef(Promise.resolve(viewport));
  const [showEncodingsPopup, setShowEncodingsPopup] = useState(false);
  const [showSearchPopup, setShowSearchPopup] = useState(false);
  const [searchTerm, setSearchTerm] = useState("");

  /** @type {(viewport: ViewerFileViewport) => Promise<ViewerFileViewport>} */
  function updated(viewport) {
    props.setViewport(viewport);
    return Promise.resolve(viewport);
  }

  /** @type {(lines: number, from?: number) => void} */
  function onMoveUp(lines, from = viewport.position) {
    readP.current = readP.current
      .then((viewport) => viewport.moveUp(lines, from))
      .then(updated);
  }

  /** @type {(lines: number) => void} */
  function onMoveDown(lines) {
    readP.current = readP.current
      .then((viewport) => viewport.moveDown(lines))
      .then(updated);
  }

  /** @type {(from?: number) => void} */
  function onReload(from = viewport.position) {
    readP.current = readP.current
      .then((viewport) => viewport.reload(from))
      .then(updated);
  }

  /** @type {() => void} */
  function onWrap() {
    readP.current = readP.current.then((viewport) => {
      const wrap = !viewport.wrap;
      const column = wrap ? 0 : viewport.column;
      return updated(viewport.updated({ wrap, column }));
    });
  }

  /** @type {(encoding: string) => void} */
  function onEncoding(encoding) {
    readP.current = readP.current.then((viewport) =>
      updated(viewport.updated({ encoding }))
    );
  }

  /** @type {(dx: number) => void} */
  function onColumn(dx) {
    readP.current = readP.current.then((viewport) => {
      const col = viewport.column + dx;
      if (col >= 0 && col < 1000) {
        return updated(viewport.updated({ column: col }));
      }

      return viewport;
    });
  }

  /** @type {(keyFull: string) => void} */
  function onKeypress(keyFull) {
    if (!props.onKeypress(keyFull)) {
      switch (keyFull) {
        case "f2":
          onWrap();
          break;
        case "f7":
          setShowSearchPopup(true);
          break;
        case "f8":
          setShowEncodingsPopup(true);
          break;
        case "left":
          onColumn(-1);
          break;
        case "right":
          onColumn(1);
          break;
        case "C-r":
          onReload();
          break;
        case "home":
          onReload(0);
          break;
        case "end":
          onMoveUp(viewport.height, viewport.size);
          break;
        case "up":
          onMoveUp(1);
          break;
        case "pageup":
          onMoveUp(viewport.height);
          break;
        case "down":
          onMoveDown(1);
          break;
        case "pagedown":
          onMoveDown(viewport.height);
          break;
      }
    }
  }

  useLayoutEffect(() => {
    readP.current = readP.current.then(() => viewport.reload()).then(updated);
  }, [
    viewport.encoding,
    viewport.size,
    viewport.width,
    viewport.height,
    viewport.wrap,
  ]);

  return h(
    viewerInput,
    {
      inputRef: props.inputRef,
      onWheel: (up) => {
        if (up) onMoveUp(1);
        else onMoveDown(1);
      },
      onKeypress,
    },
    h("text", {
      width: viewport.width,
      height: viewport.height,
      style: ViewerContent.contentStyle(theme),
      wrap: false,
      content: viewport.content,
    }),

    showEncodingsPopup
      ? h(encodingsPopup, {
          encoding: props.viewport.encoding,
          onApply: onEncoding,
          onClose: () => {
            setShowEncodingsPopup(false);
          },
        })
      : null,

    showSearchPopup
      ? h(textSearchPopup, {
          onSearch: (term) => {
            setShowSearchPopup(false);
            setSearchTerm(term);
          },
          onCancel: () => {
            setShowSearchPopup(false);
          },
        })
      : null,

    searchTerm.length > 0
      ? h(viewerSearch, {
          searchTerm,
          onComplete: () => {
            setSearchTerm("");
          },
        })
      : null
  );
};

ViewerContent.displayName = "ViewerContent";
ViewerContent.viewerInput = ViewerInput;
ViewerContent.encodingsPopup = EncodingsPopup;
ViewerContent.textSearchPopup = TextSearchPopup;
ViewerContent.viewerSearch = ViewerSearch;

/** @type {(theme: FileListTheme) => BlessedStyle} */
ViewerContent.contentStyle = (theme) => {
  const style = theme.fileList.regularItem;
  return {
    bold: style.bold,
    bg: style.bg,
    fg: style.fg,
  };
};

export default ViewerContent;
